R/prep_card.R

```diff
@@ -0,0 +1,330 @@
+#' Prepare card for tfrmt
+#'
+#' What does the preparation function do?
+#'   * `unite_data_vars()`: bring all categorical variables levels into a single
+#'     column, called `variable_level` (where applicable).
+#'  * `process_big_n()` identifies the bigN columns and changes `stat_name` to
+#'  `"bigN"`
+#'  * `process_categorical_vars()`: once we have bigN, it renames some of the
+#'  values in the categorical variable columns
+#'  * `fill_variables()`: in a hierarchical stack fills NA in one column
+#'  (with `"Any <column_name>"`) based on the presence of data in another column
+#'
+#' @inheritParams shuffle_card
+#' @param column (character) name of column(s) to use as header.
+#' @param variables (character) `cards` variables
+#'
+#' @returns a `data.frame`
+#' @export
+#'
+#' @examples
+prep_card <- function(x,
+                      column,
+                      variables = NULL,
+                      fill_overall = "Overall {colname}",
+                      fill_hierarchical_overall = "Any {colname}") {
+
+  # TODO prep_card only works with shuffled_cards
+  # TODO class the output of shuffle_card()
+  # and then shuffle if the object is not shuffled
+  # TODO check the error is propagated from the right caller env
+  # TODO check prep_card works both on its own and immediately following shuffle_card()
+
+  shuffled_card <- x
+
+  ard_args <- attr(shuffled_card, "args")
+
+  if (!is_shuffled_card(x)) {
+    shuffled_card <- shuffle_card(
+      x,
+      by = ard_args$by %||% column,
+      fill_overall = fill_overall,
+      fill_hierarchical_overall = fill_hierarchical_overall
+    )
+  }
+
+  if (!is.character(column)) {
+    cli::cli_abort(
+      "{.arg column} must be a character vector."
+    )
+  }
+
+  # TODO get the logic to work with strings and then add support for symbols /
+  # unquoted strings
+
+  # TODO priority for extracting context - e.g. by variables, etc:
+  #   1. direct passing of args
+  #   2. from attributes
+  #   3. tfrmt object
+
+  # column <- rlang::enquo(column)
+
+  # a <- tfrmt_find_args(..., env = environment(), parent_env = caller_env())
+  # columns_quo <- a$column |>
+  #   rlang::quos_auto_name()
+  # columns_char <- names(columns_quo)
+
+  if (has_attributes(x)) {
+    x <- x |>
+      # remove attributes for now
+      # TODO add some logic to deal with them
+      dplyr::filter(
+        .data$context != "attributes"
+      )
+  }
+
+  interim <- shuffled_card
+
+  # don't unite for hierarchical stack
+  if (!"hierarchical" %in% unique(x$context)) {
+    interim <- unite_data_vars(shuffled_card, column)
```

Comment:
this condition is too narrow. For example, if `ard_strata` is used to perform a calculation over multiple levels of a group, you end up with a hierarchical scenario without context = "hierarchical"

Reply:
I've moved the checking for this condition inside `unite_data_vars()` and added another one on the impact of `tidyr::unite()`. Basically we should only unite when its outcome would be identical to `dplyr::coalesce()`.